File: Algorithms/21. Merge Two Sorted Lists.rs
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
// 
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn merge_two_lists(l1: Option<Box<ListNode>>, l2: Option<Box<ListNode>>) -> Option<Box<ListNode>> {
    match (l1, l2) {
        (Some(l), None) => return Some(l),
        (None, Some(r)) => return Some(r),
        (None, None) => return None,
        (Some(l), Some(r)) => {
          if l.val <= r.val {
            return Some(Box::new(ListNode{ next: Solution::merge_two_lists(l.next, Some(r)), val: l.val }));
          } else {
            return Some(Box::new(ListNode{ next: Solution::merge_two_lists(Some(l), r.next), val: r.val }));
          }
        },
      }
    }
}

File: Algorithms/21. Merge Two Sorted Lists.ts
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function mergeTwoLists(list1: ListNode | null, list2: ListNode | null): ListNode | null {
  let result = null;
  let indicator = null;
  while (list1 || list2) {
    if (!list1) {
      if (!result) {
        result = new ListNode(list2.val);
        indicator = result;
      } else {
        indicator.next = new ListNode(list2.val);
        indicator = indicator.next;
      }
      list2 = list2.next;
      continue;
    }

    if (!list2) {
      if (!result) {
        result = new ListNode(list1.val);
        indicator = result;
      } else {
        indicator.next = new ListNode(list1.val);
        indicator = indicator.next;
      }
      list1 = list1.next;
      continue;
    }

    if (list1.val <= list2.val) {
      if (!result) {
        result = new ListNode(list1.val);
        indicator = result;
      } else {
        indicator.next = new ListNode(list1.val);
        indicator = indicator.next;
      }
      list1 = list1.next;
    } else {
      if (!result) {
        result = new ListNode(list2.val);
        indicator = result;
      } else {
        indicator.next = new ListNode(list2.val);
        indicator = indicator.next;
      }
      list2 = list2.next;
    }
  }
  return result;
};
